src/Streamly/Internal/Data/Unfold/Enumeration.hs

```diff
@@ -0,0 +1,336 @@
+enumerateFromThenToFractional stride to =
+    takeWhile (<= to + 1 / 2) $ enumerateFromStepNum stride
+
+{-# INLINE_NORMAL enumerateFromThenFractional #-}
+enumerateFromThenFractional :: (Monad m, Fractional a) => a -> Unfold m a a
+enumerateFromThenFractional = enumerateFromStepNum
+
+-------------------------------------------------------------------------------
+-- Enumerable type class
+-------------------------------------------------------------------------------
+-- | Types that can be enumerated as a stream. The operations in this type
+-- class are equivalent to those in the 'Enum' type class, except that these
+-- generate a stream instead of a list. Use the functions in
+-- "Streamly.Internal.Data.Unfold.Enumeration" module to define new instances.
+--
+-- /Pre-release/
+class Enum a => Enumerable a where
+
+-- @enumerateFrom from@ generates a stream starting with the element
+-- @from@, enumerating up to 'maxBound' when the type is 'Bounded' or
+-- generating an infinite stream when the type is not 'Bounded'.
+--
+-- @
+-- >>> Stream.toList $ Stream.take 4 $ Stream.unfold Unfold.enumerateFrom (0 :: Int)
+-- [0,1,2,3]
+--
+-- @
+--
+-- For 'Fractional' types, enumeration is numerically stable. However, no
+-- overflow or underflow checks are performed.
+--
+-- @
+-- >>> Stream.toList $ Stream.take 4 $ Stream.unfold Unfold.enumerateFrom 1.1
+-- [1.1,2.1,3.1,4.1]
+--
+-- @
+--
+-- /Pre-release/
+--
+    enumerateFrom :: (Integral a, Monad m) => Unfold m a a
+
+-- Generate a finite stream starting with the element @from@, enumerating
+-- the type up to the value @to@. If @to@ is smaller than @from@ then an
+-- empty stream is returned.
+--
+-- @
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromTo 4) 0
+-- [0,1,2,3,4]
+--
+-- @
+--
+-- For 'Fractional' types, the last element is equal to the specified @to@
+-- value after rounding to the nearest integral value.
+--
+-- @
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromTo 4) 1.1
+-- [1.1,2.1,3.1,4.1]
+--
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromTo 4.6) 1.1
+-- [1.1,2.1,3.1,4.1,5.1]
+--
+-- @
+--
+-- /Pre-release/
+    enumerateFromTo :: (Integral a, Monad m) => a -> Unfold m a a
+
+-- @enumerateFromThen from then@ generates a stream whose first element
+-- is @from@ and the successive elements are
+-- in increments of @then@.  Enumeration can occur downwards or
+-- upwards depending on whether @then@ comes before or after @from@. For
+-- 'Bounded' types the stream ends when 'maxBound' is reached, for
+-- unbounded types it keeps enumerating infinitely.
+--
+-- @
+-- >>> Stream.toList $ Stream.take 4 $ Stream.unfold (Unfold.enumerateFromThen 2) 0
+-- [0,2,4,6]
+--
+-- >>> Stream.toList $ Stream.take 4 $ Stream.unfold (Unfold.enumerateFromThen (-2)) 0
+-- [0,-2,-4,-6]
+--
+-- @
+--
+-- /Pre-release/
+    enumerateFromThen :: (Integral a, Monad m) => a -> Unfold m a a
```

Comment:
I think, the signature should be `Unfold (a, a) a`. We should move all the inputs to unfold.